lib/Sema/TypeCheckConstraints.cpp

```diff
@@ -477,6 +477,44 @@ static bool findNonMembers(ArrayRef<LookupResultEntry> lookupResults,
   return AllDeclRefs;
 }
 
+/// Find the next element in a chain of members. If this expression is (or
+/// could be) the base of such a chain, this will return \c nullptr.
+static Expr *getMemberChainSubExpr(Expr *expr) {
+  assert(expr && "getMemberChainSubExpr called with null expr!");
+  if (auto *UDE = dyn_cast<UnresolvedDotExpr>(expr)) {
+    return UDE->getBase();
+  } else if (auto *CE = dyn_cast<CallExpr>(expr)) {
+    return CE->getFn();
+  } else if (auto *BOE = dyn_cast<BindOptionalExpr>(expr)) {
+    return BOE->getSubExpr();
+  } else if (auto *FVE = dyn_cast<ForceValueExpr>(expr)) {
+    return FVE->getSubExpr();
+  } else if (auto *SE = dyn_cast<SubscriptExpr>(expr)) {
+    return SE->getBase();
+  } else {
```

Comment:
@rintaro pointed out that there is a problem with code completion in the chains e.g. `.bar.#^HERE^#` I believe that all we need to do is add another `else if` here which returns a base expression from `CodeCompletionExpr` @Jumhyn could you please add that?

Reply:
Just FYI, for
```
let a: Foo = .bar.<HERE>
```
the parsed initializer would be like
```
(code_completion_expr
  (unresolved_member_expr name="bar"))
```

Reply:
It would be great if you could add  some code completion test cases including:
```
struct Foo {
  static var bar = Bar()
}
struct Bar {
  var anotherFoo = Foo()
}
let a: Foo = .bar.#^HERE^#
```

Reply:
Ooh, good catch @rintaro—I had played around with code completion for homogenously-typed chains, but I guess I didn't test heterogeneously-typed chains. Thank you! Will update and add appropriate tests next chance I get.

Also after @benlangmuir's comment on the review thread and playing with the toolchain just now I'm thinking more about code completion and I'm interested to hear what (all of) you think: should we offer (de-prioritized) completion results for `UnresolvedMemberExpr`s which don't match the contextual type? I.e., in the above example for a completion:
```
let a: Foo = .#^HERE^#
```
should we offer `.bar` since there are no static members of type `Foo` (or, if there were such static members, should we offer `.bar` *after* all the `Foo`-typed members)?

Reply:
@rintaro Fixed code completion according to @xedin's suggestion and added some tests in _test/IDE/complete_unresolved_chains.swift_. Let me know if you have suggestions for additional tests!

Reply:
> if there were such static members, should we offer .bar after all the Foo-typed members.

I think that's the most reasonable solution because, if I understand correctly, contextual type is usually considered to be just a hint since it's possible to continue typing and reach `Foo` again deeper in the chain e.g. even if I pick `.bar` after I complete again and reach `Foo` though `anotherFoo`.

/cc @nathawes

Reply:
Is there a good way to express with `FileCheck` that two `CHECK-DAG` blocks should be ordered with respect to one another, but not necessarily within each block? I.e., to express that all the results matching the contextual type should come before the non-matching results?

Reply:
I think what you want is `CHECK-NEXT`?

Reply:
Hmm, what would be the matched string supplied to `CHECK-NEXT`? My impression was that we're using `CHECK-DAG` because we don't necessarily guarantee the order of returned completions in these instances, and it seems like using `CHECK-NEXT` would be saying "this *particular* non-correctly-typed completion must appear immediately after all the correctly-typed completions" rather than "all the correctly-typed completions (in any order) must appear before all the non-correctly-typed completions (in any order)".

How could I use `CHECK-NEXT` to match:
```
1
2
3
4
```
or
```
2
1
4
3
```
but not
```
1
3
2
4
```
?
The best thing I can come up with off the top of my head is something like
```
CHECK-DAG: 1
CHECK-DAG: 2
CHECK-NOT: string_that_will_never_occur
CHECK-DAG: 3
CHECK-DAG: 4
```
but I'm curious if there's a better way...

Reply:
Actually, is testing the ordering here the right approach here at all? What is the appropriate mechanism for prioritizing (or deprioritizing) code completion results? cc @rintaro @benlangmuir

Reply:
Sorting the result is not a responsibility of `libIDE`. So you don't need the ordering for `swift-ide-test` test cases. Clients (e.g. `SourceKit`) sort the result using `ExprSpecific` flag and `TypeRelation`.

Reply:
Having `ExprSpecific` have highest priority regardless of `TypeRelation`. For example
```
func foo(x: Enum1, y: Int) {}
func foo(x: Enum2, y: String) {}

foo(x: .#^HERE^#, y: "")
```
In this case, cases from `Enum2` have `ExprSpecific` and `TypeRelation[Identical]`, but cases from `Enum1` only have `TypeRelation[Identical]`.

Reply:
Awesome, thanks for the context on that. So from my understanding, I should just make sure that the non-matching results are present, and that the matching results have `ExprSpecific` and `TypeRelation` set appropriately? (Also, as an aside, is there a reason that we set the `ExprSpecific` flag for static var, static func, and enum members but not for constructor members, or is that just an oversight?)

Reply:
> I should just make sure that the non-matching results are present, and that the matching results have ExprSpecific and TypeRelation set appropriately? 

Right.
Also `TypeRelation[Invalid]` for static methods returning `Void` (This should be done automatically.)

> is there a reason that we set the ExprSpecific flag for static var, static func, and enum members but not for constructor members?

I actually don't remember. But let's leave it for now :)

Reply:
(We still want to suggest `Void` members because the user can modify it to return something)

Reply:
Makes sense, thanks @rintaro!

Reply:
> is there a reason that we set the ExprSpecific flag for static var, static func, and enum members but not for constructor members?

We want to prioritize enum-like values over constructors.  While it is legal to use `.init(...)` in many places, usually when you type a leading `.` you are looking for an enum-like static value.